articles/cross_compilation_build_tools_design.md

```diff
@@ -0,0 +1,248 @@
+---
+layout: default
+title: Cross-compiling ROS 2 packages
+abstract:
+  This article is a design proposal for extending the build tools for ROS 2 to
+  simplify cross-compiling ROS 2 packages for platforms currently not supported
+  by ROS 2, or building on a host with different platforms than the target
+  platform. That later use case is particularly useful for speeding up build
+  times by building the code on a powerful development workstation, and then
+  deploying the binaries to a constrained resources hardware that uses the
+  target platform, thus making the development cycle much faster.
+author: >
+  [Thomas Moulard](https://github.com/thomas-moulard),
+  [Juan Rodriguez Hortala](https://github.com/juanrh)
+published: false
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Background
+
+The [cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation)
+contains instructions for cross-compiling ROS 2 to unsupported platforms, using
+some configuration files for CMake and Docker that is setting maintained in the
+[ros2/cross_compile repository](https://github.com/ros2/cross_compile/).
+However, that tutorial doesn't provide instructions to build other ROS 2
+packages, but just for building a whole ROS 2 distribution. Ideally, we should
+be able to cross-compile a ROS package for ROS 2 by launching a single
+`colcon build` command. Finally, cross-compiling for other platforms besides
+ARM64 is not supported in the [ros2/cross_compile repository](https://github.com/ros2/cross_compile/).
+
```

Comment:
Worth mentioning is https://github.com/ruslo/polly, which contains a bunch of toolchains for crosscompiling CMake projects for a variety of platforms.

Reply:
Sounds awesome, thanks!

Reply:
armhf is also supported with the [ros2/cross_compile repository](https://github.com/ros2/cross_compile/). The architecture is an argument of the "sysroot dockerfile"

> Use an argument to specify the architecture of the target-file system.
This can be specified with the `--build-arg` argument on the docker
build command:
  *for armhf: docker build --build-arg ARM_ARCH=arm32v7 ...
  *for aarch64: docker build --build-arg ARM_ARCH=arm64v8 ...

Reply:
[Dockerfile_ubuntu_arm](https://github.com/ros2/cross_compile/blob/master/sysroot/Dockerfile_ubuntu_arm) indeed supports that argument, but [entry_point.sh](https://github.com/ros2/cross_compile/blob/master/entry_point.sh#L28) assumes ARM64 when it sets `export TARGET_TRIPLE=aarch64-linux-gnu`. However I agree we could be more precise and state that the tutorial does support ARMHF, but the automation described in the "Automated Cross-compilation" section does not fully support it. I'll be more precise about that, as this proposal doesn't try to change the basic approach to cross compilation, but to pack it into a development tool that is easy to use